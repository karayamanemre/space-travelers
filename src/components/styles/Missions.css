table {
  margin: 25px 100px;
  border-collapse: collapse;
}

table,
td,
th {
  border: 2px solid #ddd;
  padding: 15px;
}

tr:nth-child(even) {
  background-color: #f2f2f2;
}

button {
  font-size: 1rem;
  width: 100px;
  height: 50px;
}

.status {
  font-size: 1rem;
}
