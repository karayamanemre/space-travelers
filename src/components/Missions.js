import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import styles from './styles/Missions.module.css';
import { fetchMissions, joinMission, leaveMission } from '../redux/missions/missions';

const Missions = () => {
  const missions = useSelector((state) => state.missions);
  const dispatch = useDispatch();
  useEffect(() => {
    if ((missions || []).length === 0) {
      dispatch(fetchMissions());
    }
  }, [dispatch, missions]);
  const toggleReservation = (mission) => {
    if (!mission.reserved) {
      dispatch(joinMission(mission.id));
    } else {
      dispatch(leaveMission(mission.id));
    }
  };
  const missionHeadings = ['Missions', 'Description', 'Status', 'Action'];
  return (
    <table className={styles.mission_table}>
      <thead>
        <tr>
          {missionHeadings.map((heading) => (
            <th className={styles.mission_th} key={heading}>{heading}</th>
          ))}
        </tr>
      </thead>
      <tbody>
        {missions.map((mission) => (
          <tr key={mission.id}>
            <td className={styles.mission_td}><b>{mission.mission_name}</b></td>
            <td className={styles.mission_td}>{mission.description}</td>
            <td className={styles.mission_td}>
              <div className="status">
                {mission.reserved && 'Active'}
                {!mission.reserved && 'Not a Member'}
              </div>
            </td>
            <td className={styles.mission_td}>
              <button
                key={mission.id}
                type="button"
                onClick={() => toggleReservation(mission)}
              >
                {mission.reserved && 'Leave Mission'}
                {!mission.reserved && 'Join Mission'}
              </button>
            </td>
          </tr>
        ))}
      </tbody>
    </table>
  );
};

export default Missions;
