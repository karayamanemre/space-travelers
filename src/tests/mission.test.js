import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../redux/configureStore';
import Missions from '../components/Missions';
import missionsReducer, { showMissions, joinMission } from '../redux/missions/missions';
import missionsData from './missionData';

describe('Mission', () => {
  it('Create mission item', () => {
    render(
      <Provider store={store}>
        <Missions />
      </Provider>,
    );
    expect(Missions).toMatchSnapshot();
  });

  it('Update the state with API data', () => {
    expect(missionsReducer(missionsData, showMissions(missionsData))).toEqual(
      missionsData,
    );
  });

  it('Update the state with correct id to reserved: true', () => {
    const missionId = '8E0A6D0';
    expect(missionsReducer(missionsData, joinMission(missionId))).toEqual(
      missionsData.map((item) => (item.id === missionId ? { ...item, reserved: true } : item)),
    );
  });
});
