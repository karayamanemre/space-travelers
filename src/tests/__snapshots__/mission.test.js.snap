// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Create mission item 1`] = `[Function]`;
